### PR TITLE
SPEC2017 integration

### PR DESCRIPTION
**Pull request by hnpl**
Mostly identical to SPEC 2006 tutorial. The scripts are being updated.

### CODE REVIEW

File: docs/spec2017-tutorial.md [hunk 1]
```diff
@@ -0,0 +1,586 @@
+# Tutorial: Run SPEC CPU 2017 Benchmarks in Full System Mode with gem5art  
+
+## Introduction  
+In this tutorial, we will demonstrate how to utilize gem5art to run SPEC CPU 2017 benchmarks in gem5 full system mode. 
+The full example with all of the gem5art tutorials can be found [here](https://github.com/darchr/gem5art-experiments). 
+The scripts in this tutorial work with gem5art-* v0.3.1.  
+
+**Note**: This steps in this tutorial are mostly identical to that of [the SPEC 2006 tutorial](spec2006-tutorial.md).
```
Comment (powerjg) [inline docs/spec2017-tutorial.md]: Do you think it would be better to *not* replicate the text here and instead just say "Here's what needs to change"?

**Review by powerjg — COMMENTED** [review]
Looking great so far! One comment below. Let me know what you think.

File: docs/spec2017-tutorial.md (hunk 1 above)
Comment (hnpl) [reply docs/spec2017-tutorial.md]: I would expect someone to read SPEC 2017 tutorial without reading the other one. I'll point to the differences at the beginning of this tutorial.

**Comment by hnpl** [comment]
Hmmm, none of the spec2017 benchmarks works.
I also tried running the benchmarks using qemu, none worked.

**Comment by powerjg** [comment]
What a about natively?

I would guess there's some config problem.

**Comment by hnpl** [comment]
Fixed the illegal instructions problem. Squashing and merging.

**Comment by hnpl** [comment]
When we compile the benchmark, we use packer, which exposes the information of the host CPU to the guest. We shouldn't use `-march=native` as there will be lots of the host CPU's instructions are not available on the guest CPU.

File: docs/spec2006-tutorial.md [hunk 2]
```diff
@@ -598,7 +597,7 @@ TODO
 [5] J. L. Hennessy and D. A. Patterson, “A new golden age for computer architecture,” *Communications of the ACM*, vol. 62, no. 2, pp. 48–60, 2019.  
  
 
-## Apprendix I. Working SPEC 2006 Benchmarks x CPU Model matrix
+## Apprendix I. Working SPEC 2006 Benchmarks x CPU Model Matrix
```
Comment (powerjg) [inline docs/spec2006-tutorial.md:600]: Spelling: Apprendix -> Appendix

File: docs/spec2017-tutorial.md [hunk 3]
```diff
@@ -0,0 +1,635 @@
+# Tutorial: Run SPEC CPU 2017 Benchmarks in Full System Mode with gem5art  
+
+## Introduction  
+In this tutorial, we will demonstrate how to utilize gem5art to run SPEC CPU 2017 benchmarks in gem5 full system mode. 
+The full example with all of the gem5art tutorials can be found [here](https://github.com/darchr/gem5art-experiments). 
+The scripts in this tutorial work with gem5art-* v0.3.1.
```
Comment (powerjg) [inline docs/spec2017-tutorial.md:6]: Are you sure this shouldn't be 0.4.1?

File: docs/spec2017-tutorial.md [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+In the root folder of the experiment,
+
+```sh
+python3 launch_spec2017_experiment.py
+```
+
+## Getting the Results  
+TODO
+
+## Apprendix I. Working SPEC 2017 Benchmarks x CPU Model Matrix
```
Comment (powerjg) [inline docs/spec2017-tutorial.md]: Spelling :)

**Review by powerjg — COMMENTED** [review]
A couple of minor comments below. Otherwise, we should merge this and then continue working on it.

File: docs/spec2017-tutorial.md (hunk 3 above)
Comment (hnpl) [reply docs/spec2017-tutorial.md:6]: I haven't updated the launch script. I'll do that later.

**Comment by hnpl** [comment]
Currently, the workloads work with KVM and all workloads do not work with O3CPU and AtomicCPU. I suspect it's due to a gem5 bug, as gem5 stopped because of an assertion failure. All workloads share this error,
`gem5.opt: build/X86/dev/x86/i8237.cc:39: virtual Tick X86ISA::I8237::read(PacketPtr): Assertion 'pkt->getSize() == 1' failed.`

File: docs/spec2017-tutorial.md (hunk 4 above)
Comment (hnpl) [reply docs/spec2017-tutorial.md]: Fixed.

**Comment by powerjg** [comment]
Good to know! That looks like a device issue... We'll have to dig into that more later.

File: docs/spec2006-tutorial.md (hunk 2 above)
Comment (hnpl) [reply docs/spec2006-tutorial.md:600]: This should be fixed when this branch is merged to the master branch.

**Review by powerjg — APPROVED** [review]
Looks good to me. Just make sure to add any outstanding things we need to do to Jira (e.g., figure out why it's failing ;))

**Comment by hnpl** [comment]
Do we need to run both speed and rate workloads in SPEC 2017? [The definitions of speed and rate are different in SPEC 2017](https://www.spec.org/cpu2017/Docs/runcpu.html#rate).  
For each workload, there are two versions: the speed version (having the name with the suffix "_s") is compiled with OpenMP directives, while the rate version (having the name with the suffix "_r") isn't.  The latter is also run with bigger datasets.

**Comment by hnpl** [comment]
More about the speed version: there is an option to specify the number of OpenMP Threads, and it is default to 4. This number *seems* arbitrary to me, should we change this?

**Comment by powerjg** [comment]
For now, let's disable openmp (https://www.spec.org/cpu2017/Docs/config.html#openmp) and only run the "speed" versions.

Remind me to discuss this more when we have our first gem5 meeting. 

Edit: Maybe we should go ahead and run both the speed and rate. We can see if it makes a difference :)

**Comment by hnpl** [comment]
I'd keep both.

**Comment by hnpl** [comment]
Merge in #62.

**Pull request closed without merging**